### ISSUE TITLE
[Docs]: How to write isolated playwright tests against a real database

### ISSUE THREAD

**Issue by nodefir** [microsoft/playwright#33699]
### Page(s)

https://playwright.dev/docs/next/browser-contexts

### Description

When writing playwright tests against a local database (i.e. a postgres server with a database named `your_app_test`), there does not seem to be anywhere online that walks one through writing transactional tests that will not leak data during parallelization. All I have been able to come across is information on fixture data, but what about data created by the test itself?

I have gone down the rabbit hole of trying to recreate the seamless system testing (end-to-end testing) provided by Rails, which is able to spin up a test server and run tests in parallel without having to worry about data collisions.

Approaches I have tried thus far:

- data truncation in before/after hooks
- data truncation around `await use()` in fixtures
- beginning and rolling back transactions around test (using `beforeEach`, `afterEach`, and then a fixture)
- creating isolated databases using something like execSync(`psql -c "CREATE DATABASE your_app_test_${env.process.TEST_WORKER_INDEX};` (doesn't work because `env.process.TEST_WORKER_INDEX` isn't available to the web server).
- data truncation using a single worker (too slow to realistically be an option - but works)

How can I achieve a similar workflow to Rails system tests, where I can create data (directly in the test or by clicking around in my app) in a test database without worrying about other tests creating similar data? Is this out of scope for playwright?

For what it's worth, I am using prisma and Next.js.

**Comment by dgozman** [comment]
@nodefir Thank you for the issue! This is what we'd recommend for you:
- loosely follow [this guide](https://playwright.dev/docs/test-parallel#isolate-test-data-between-parallel-workers);
- instead of running your `Next.js` server as a `webServer`, start it in the worker fixture from the guide above;
- make sure your `Next.js` server creates a unique database in this setup;
- you might want to override `baseURL` fixture to use the freshly launched server's address;
- teardown your `Next.js` server in the fixture cleanup.

This will ensure that tests running in one worker will talk to its own server instance that will connect to its own database. Let me know what you think.

**Comment by nodefir** [comment]
Thanks @dgozman! I'll take a look this evening. If I can get this working I'll write up a blog post and also post the solution here.

**Comment by mxschmitt** [comment]
I'll close it for now since it seems answered feel free to re-file if it doesn't work for you.

**Comment by northfir** [comment]
I'm the original author of this issue, but wanted to come back after finally having time to mess with it again. I built: https://github.com/testdrivendev/iodome to solve this problem for myself. It should get people up and running or at least provide an example of how to set this up with Next.js, postgres (just prisma support for now), and playwright.
